File: 2016/src/bin/01.rs
use std::collections::HashSet;

use aoc2016::aoc;


pub fn part1(input:&str) -> String {
    let instrs = input.split(", ");
    let mut x = 0; 
    let mut y = 0;
    let mut idx = 0;
    let dirs = [(0,-1),(1,0),(0,1),(-1,0)];
    for instr in instrs {
        let dir = instr.chars().nth(0).unwrap();
        if dir == 'R' {
            idx += 1;
        }
        else {
            idx += 3;
        }
        idx %= 4;
        let dist = &instr[1..].parse::<i32>().unwrap();
        x += dist * dirs[idx].0;
        y += dist * dirs[idx].1;
    }
    return (x.abs() + y.abs()).to_string();
}

pub fn part2(input:&str) -> String {
    let instrs = input.split(", ");
    let mut x: i32 = 0;
    let mut y: i32 = 0;
    let mut idx = 0;
    let dirs = [(0,-1),(1,0),(0,1),(-1,0)];
    let mut set = HashSet::new();

    for instr in instrs {
        let dir = instr.chars().nth(0).unwrap();
        if dir == 'R' {
            idx += 1;
        }
        else {
            idx += 3;
        }
        idx %= 4;

        let dist = (&instr[1..]).parse::<i32>().unwrap();
        for _ in 0..dist {

            set.insert((x,y));
            x += dirs[idx].0;
            y += dirs[idx].1;
            if set.contains(&(x,y)) {
                return (x.abs() + y.abs()).to_string();
            }
        }
    }
    unreachable!();
}

fn main(){
    aoc::run(1,&part1,&part2);
}

File: 2025/01.py
from utils import run

def part1(input: str):
    input = input.strip().split("\n")
    n = 50
    count = 0
    for line in input:
        sign = 1
        if line[0] == "L":
            sign =  -1
            
        c = line[1:]
        num = int(c) * sign
        
        n = (n + num) % 100
        if n == 0:
            count += 1
    return count

def part2(input: str):
    input = input.strip().split("\n")
    n = 50
    count = 0
    for line in input:
        sign = 1
        if line[0] == "L":
            sign =  -1
            
        c = line[1:]        
        for i in range(int(c)):
            n = (n + sign) % 100
            if n == 0:
                count += 1
        
    return count


run(1, part1, part2)